dlab_core/infrastructure/services.py

```diff
@@ -0,0 +1,158 @@
+
+
+class KeyCloakError(DLabException):
+    pass
+
+
+class InvalidParameterFormatError(KeyCloakError):
+    message = 'Parameter {0} is not of {1} type: {2}'
+
+    def __str__(self):
+        return self.message.format(*self.args)
+
+
+class KeyCloakConnectionError(KeyCloakError):
+    pass
+
+
+class KeyCloak(object):
+    """Client for KeyCloak interaction
+
+    :type keycloak_host: str
+    :param keycloak_host: IP or hostname for KeyCloak server to use
+
+    :type realm_name: str
+    :param realm_name: name of the KeyCloak Realm to use
+
+    :type client_id: str
+    :param client_id: KeyCloak Client to use for token introspection
+
+    :type client_secret: str
+    :param client_secret: KeyCloak Client secret string
+    """
+    def __init__(self, keycloak_host, realm_name, client_id, client_secret):
+        check = ['keycloak_host', 'realm_name', 'client_id', 'client_secret']
+        for arg_name in check:
+            arg_value = locals()[arg_name]
+            if not isinstance(arg_value, str):
+                raise InvalidParameterFormatError(
+                    arg_name, 'str', type(arg_value).__name__
+                )
+        self._realm_address = '{}/realms/{}'.format(
+            keycloak_host.rstrip('/'), realm_name
+        )
+        self._client_id = client_id
+        self._client_secret = client_secret
+        self._pub_key = self.get_key()
+
+    def make_request_to_server(self, method='GET', endpoint='', **kwargs):
+        """Performs request to endpoint
+
+        :type method: str
+        :param method: method for the HTTP request
+        :type endpoint: str
+        :param endpoint: endpoint to make request to
+
+        :rtype: requests.Response
+        :return: HTTP response object
+        """
+        kwargs['verify'] = kwargs.get('verify') or False
+        try:
+            response = requests.request(method, endpoint, **kwargs)
+            response.raise_for_status()
+            return response
+        except requests.exceptions.RequestException as e:
+            raise KeyCloakConnectionError(str(e))
+
+    def get_key(self):
+        """Retrieves public key from KeyCloak realm
+
+        :rtype: str
+        :return: public key for token decoding
+        """
+        response = self.make_request_to_server(endpoint=self._realm_address)
+        if response:
+            # PyJWT expects key to be in PEM format (including header/footer)
+            return (
+                '-----BEGIN PUBLIC KEY-----\n{}'
+                '\n-----END PUBLIC KEY-----'
+            ).format(response.json().get('public_key'))
+
+    def validate_token(self, access_token):
+        """Validates signature of given token and introspects it
+
+        :type access_token: str
+        :param access_token: JWT-token
+
+        :rtype: bool
+        :return: True if token is valid and active, False otherwise
+        """
+        self._pub_key = self._pub_key or self.get_key()
```

Comment:
you can use only if here
you are not using self._pub_key here why do you have this line here